File: medium/3660-jump-game-ix/solution.py
from typing import List


class Solution:
    def maxValue(self, nums: List[int]) -> List[int]:
        """
        3660. Jump Game IX
        Time: O(n)
        Space: O(n)
        """
        n = len(nums)
        suffix_min = nums[:]

        for i in range(n - 2, -1, -1):
            suffix_min[i] = min(nums[i], suffix_min[i + 1])

        ans = [0] * n
        start = 0
        component_max = nums[0]

        for i, num in enumerate(nums):
            component_max = max(component_max, num)

            if i == n - 1 or component_max <= suffix_min[i + 1]:
                for j in range(start, i + 1):
                    ans[j] = component_max

                if i + 1 < n:
                    start = i + 1
                    component_max = nums[start]

        return ans

File: medium/3660-jump-game-ix/solution.java
/**
 * 3660. Jump Game IX
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public int[] maxValue(int[] nums) {
        int n = nums.length;
        int[] suffixMin = new int[n];
        suffixMin[n - 1] = nums[n - 1];

        for (int i = n - 2; i >= 0; i--) {
            suffixMin[i] = Math.min(nums[i], suffixMin[i + 1]);
        }

        int[] ans = new int[n];
        int start = 0;
        int componentMax = nums[0];

        for (int i = 0; i < n; i++) {
            componentMax = Math.max(componentMax, nums[i]);

            if (i == n - 1 || componentMax <= suffixMin[i + 1]) {
                for (int j = start; j <= i; j++) {
                    ans[j] = componentMax;
                }

                if (i + 1 < n) {
                    start = i + 1;
                    componentMax = nums[start];
                }
            }
        }

        return ans;
    }
}
